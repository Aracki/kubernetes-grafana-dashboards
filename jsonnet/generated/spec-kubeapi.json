{
   "rows": [
      {
         "panels": [
            {
               "formula": "sum by (verb, code)(\n    rate(apiserver_request_count{verb!~\"$verb_excl\", code=~\"5..\"}[5m])\n    ) / ignoring(code) group_left\nsum by (verb)(\n    rate(apiserver_request_count[5m])\n    )\n",
               "legend": "{{ verb }} - {{ code }}",
               "threshold": 0.01,
               "title": "API Error ratio 500s/total (except $verb_excl)"
            }
         ],
         "title": "API Error rate"
      },
      {
         "panels": [
            {
               "formula": "histogram_quantile (\n    0.$api_percentile,\n    sum by (le, verb)(\n      rate(apiserver_request_latencies_bucket{verb!~\"$verb_excl\"}[5m])\n      )\n    ) / 1e3 > 0\n",
               "legend": "{{ verb }}",
               "threshold": 200,
               "title": "API $api_percentile-th latency[ms] by verb (except $verb_excl)"
            }
         ],
         "title": "API Latency"
      },
      {
         "panels": [
            {
               "formula": "sum by (operation, instance)(\n  rate(etcd_request_latencies_summary{job=\"kubernetes_apiservers\",quantile=\"0.9\"}[5m]) < Inf\n)/ 1e3\n",
               "legend": "{{ instance }} - {{ operation }}",
               "threshold": 20,
               "title": "etcd 90th latency[ms] by (operation, instance)"
            }
         ],
         "title": "etcd Latency"
      }
   ],
   "templates_custom": [
      {
         "default": "90",
         "hide": "",
         "name": "api_percentile",
         "values": "50, 90, 99"
      },
      {
         "default": "(CONNECT|WATCH)",
         "hide": "variable",
         "name": "verb_excl",
         "values": "(CONNECT|WATCH)"
      }
   ]
}
