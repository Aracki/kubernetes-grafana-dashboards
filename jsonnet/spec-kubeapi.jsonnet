{
  local runbook_url = 'https://engineering-handbook.nami.run/sre/runbooks/kubeapi',
  templates_custom: {
    api_percentile: {
      values: '50, 90, 99',
      default: '90',
      hide: '',
    },
    verb_excl: {
      values: '(CONNECT|WATCH)',
      default: '(CONNECT|WATCH)',
      hide: 'variable',
    },
  },
  rows: [
    local t = $.templates_custom;
    {
      title: 'Kube API',
      panels: [
        {
          local this = self,
          title: 'API Error ratio 500s/total (except $verb_excl)',
          formula: |||
            sum by (verb, code)(
              rate(apiserver_request_count{verb!~"$verb_excl", code=~"5.."}[5m])
            ) / ignoring(code) group_left
            sum by (verb)(
              rate(apiserver_request_count[5m])
            )
          |||,
          legend: '{{ verb }} - {{ code }}',
          threshold: 0.01,
          alert: 'KubeAPIErrorRatioHigh',
          alert_expr: |||
            sum by (instance)(
              rate(apiserver_request_count{verb!~"%s", code=~"5.."}[5m])
            ) /
            sum by (instance)(
              rate(apiserver_request_count[5m])
            ) > %s
          ||| % [t.verb_excl.default, this.threshold],
          annotations: {
            summary: 'Kube API 500s ratio is High',
            description: |||
              Issue: Kube API Error ratio on {{ $labels.instance }} is above %s: {{ $value }}
              Playbook: %s#%s
            ||| % [this.threshold, runbook_url, this.alert],
          },
        },

        {
          local this = self,
          title: 'API $api_percentile-th latency[ms] by verb (except $verb_excl)',
          formula: |||
            histogram_quantile (
              0.$api_percentile,
              sum by (le, verb)(
                rate(apiserver_request_latencies_bucket{verb!~"$verb_excl"}[5m])
              )
            ) / 1e3 > 0
          |||,
          legend: '{{ verb }}',
          threshold: 200,
          alert: 'KubeAPILatencyHigh',
          alert_expr: |||
            histogram_quantile (
              0.%s,
              sum by (le, instance)(
                rate(apiserver_request_latencies_bucket{verb!~"%s"}[5m])
              )
            ) / 1e3 > %s
          ||| % [t.api_percentile.default, t.verb_excl.default, this.threshold],
          annotations: {
            summary: 'Kube API Latency is High',
            description: |||
              Issue: Kube API Latency on {{ $labels.instance }} is above %s: {{ $value }}
              Playbook: %s#%s
            ||| % [this.threshold, runbook_url, this.alert],
          },
        },
      ],
    },
    {
      title: 'Kube Control Manager',
      panels: [
        {
          local this = self,
          title: '',
          formula: |||
            sum by (instance)(
              APIServiceRegistrationController_work_duration{quantile="0.9"}
            )
          |||,
          legend: '{{ instance }}',
          threshold: 100,
          alert: 'KubeControllerWorkDurationHigh',
          alert_expr: |||
            sum by (instance)(
              APIServiceRegistrationController_work_duration{quantile="0.9"}
            ) > %s
          ||| % [this.threshold],
          annotations: {
            summary: 'Kube Control Manager workqueue processing is slow',
            description: |||
              Issue: Kube Control Manager on {{ $labels.instance }} work duration is above %s: {{ $value }}
              Playbook: %s#%s
            ||| % [this.threshold, runbook_url, this.alert],
          },
        },
      ],
    },
    {
      title: 'Kube Etcd',
      panels: [
        {
          local this = self,
          title: 'etcd 90th latency[ms] by (operation, instance)',
          formula: |||
            max by (operation, instance)(
              rate(etcd_request_latencies_summary{job="kubernetes_apiservers",quantile="0.9"}[5m]) < Inf
            )/ 1e3
          |||,
          legend: '{{ instance }} - {{ operation }}',
          threshold: 20,
          alert: 'KubeEtcdLatencyHigh',
          alert_expr: |||
            max by (instance)(
              rate(etcd_request_latencies_summary{job="kubernetes_apiservers",quantile="0.9"}[5m]) < Inf
            )/ 1e3 > %s
          ||| % [this.threshold],
          annotations: {
            summary: 'Etcd Latency is High',
            description: |||
              Issue: Kube Etcd latency on {{ $labels.instance }} above %s: {{ $value }}
              Playbook: %s#%s
            ||| % [this.threshold, runbook_url, this.alert],
          },
        },
      ],
    },
  ],
}
