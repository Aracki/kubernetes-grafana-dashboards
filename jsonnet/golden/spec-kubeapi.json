{
   "rows": [
      {
         "panels": [
            {
               "alert": "KubeAPIErrorRatioHigh",
               "alert_expr": "sum by (instance)(\n  rate(apiserver_request_count{verb!~\"(CONNECT|WATCH)\", code=~\"5..\"}[5m])\n) /\nsum by (instance)(\n  rate(apiserver_request_count[5m])\n) > 0.01\n",
               "annotations": {
                  "description": "Issue: Kube API Error ratio on {{ $labels.instance }} is above 0.01: {{ $value }}\nPlaybook: https://engineering-handbook.nami.run/sre/runbooks/kubeapi#KubeAPIErrorRatioHigh\n",
                  "summary": "Kube API 500s ratio is High"
               },
               "formula": "sum by (verb, code)(\n  rate(apiserver_request_count{verb!~\"$verb_excl\", code=~\"5..\"}[5m])\n) / ignoring(code) group_left\nsum by (verb)(\n  rate(apiserver_request_count[5m])\n)\n",
               "legend": "{{ verb }} - {{ code }}",
               "threshold": 0.01,
               "title": "API Error ratio 500s/total (except $verb_excl)"
            },
            {
               "alert": "KubeAPILatencyHigh",
               "alert_expr": "histogram_quantile (\n  0.90,\n  sum by (le, instance)(\n    rate(apiserver_request_latencies_bucket{verb!~\"(CONNECT|WATCH)\"}[5m])\n  )\n) / 1e3 > 200\n",
               "annotations": {
                  "description": "Issue: Kube API Latency on {{ $labels.instance }} is above 200: {{ $value }}\nPlaybook: https://engineering-handbook.nami.run/sre/runbooks/kubeapi#KubeAPILatencyHigh\n",
                  "summary": "Kube API Latency is High"
               },
               "formula": "histogram_quantile (\n  0.$api_percentile,\n  sum by (le, verb)(\n    rate(apiserver_request_latencies_bucket{verb!~\"$verb_excl\"}[5m])\n  )\n) / 1e3 > 0\n",
               "legend": "{{ verb }}",
               "threshold": 200,
               "title": "API $api_percentile-th latency[ms] by verb (except $verb_excl)"
            }
         ],
         "title": "Kube API"
      },
      {
         "panels": [
            {
               "alert": "KubeControllerWorkDurationHigh",
               "alert_expr": "sum by (instance)(\n  APIServiceRegistrationController_work_duration{quantile=\"0.9\"}\n) > 100\n",
               "annotations": {
                  "description": "Issue: Kube Control Manager on {{ $labels.instance }} work duration is above 100: {{ $value }}\nPlaybook: https://engineering-handbook.nami.run/sre/runbooks/kubeapi#KubeControllerWorkDurationHigh\n",
                  "summary": "Kube Control Manager workqueue processing is slow"
               },
               "formula": "sum by (instance)(\n  APIServiceRegistrationController_work_duration{quantile=\"0.9\"}\n)\n",
               "legend": "{{ instance }}",
               "threshold": 100,
               "title": "Kube Control Manager work duration"
            }
         ],
         "title": "Kube Control Manager"
      },
      {
         "panels": [
            {
               "alert": "KubeEtcdLatencyHigh",
               "alert_expr": "max by (instance)(\n  rate(etcd_request_latencies_summary{job=\"kubernetes_apiservers\",quantile=\"0.9\"}[5m]) < Inf\n)/ 1e3 > 20\n",
               "annotations": {
                  "description": "Issue: Kube Etcd latency on {{ $labels.instance }} above 20: {{ $value }}\nPlaybook: https://engineering-handbook.nami.run/sre/runbooks/kubeapi#KubeEtcdLatencyHigh\n",
                  "summary": "Etcd Latency is High"
               },
               "formula": "max by (operation, instance)(\n  rate(etcd_request_latencies_summary{job=\"kubernetes_apiservers\",quantile=\"0.9\"}[5m]) < Inf\n)/ 1e3\n",
               "legend": "{{ instance }} - {{ operation }}",
               "threshold": 20,
               "title": "etcd 90th latency[ms] by (operation, instance)"
            }
         ],
         "title": "Kube Etcd"
      }
   ],
   "templates_custom": {
      "api_percentile": {
         "default": "90",
         "hide": "",
         "values": "50, 90, 99"
      },
      "verb_excl": {
         "default": "(CONNECT|WATCH)",
         "hide": "variable",
         "values": "(CONNECT|WATCH)"
      }
   }
}
